pkg/cloud/rgraph/exec/retrieable_action_test.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2024 Google LLC
+
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
+)
+
+// fakeAction will return error for n actions defined in errorRunThreshold,
+// runCtr counts all action executions.
+// errorRunThreshold set to -1 means that Action should always return error.
+type fakeAction struct {
+	runCtr            int
+	errorRunThreshold int
+}
+
+func (fa *fakeAction) CanRun() bool {
+	return true
+}
+
+func (fa *fakeAction) Signal(e Event) bool {
+	return false
+}
+
+func (fa *fakeAction) Run(ctx context.Context, c cloud.Cloud) (EventList, error) {
+	fa.runCtr++
+	if fa.errorRunThreshold > fa.runCtr {
+		return EventList{}, fmt.Errorf("Action in error")
+	}
+	return EventList{}, nil
+}
+
+func (fa *fakeAction) DryRun() EventList {
+	return EventList{}
+}
+
+func (fa *fakeAction) String() string {
+	return "fakeAction"
+}
+
+func (fa *fakeAction) PendingEvents() EventList {
+	return EventList{}
+}
+
+func (fa *fakeAction) Metadata() *ActionMetadata {
+	return &ActionMetadata{
+		Name: "fakeAction",
+	}
+}
+
+// fakeRetryProvider implements RetryProvider.
+// ctr - counts all retry provider calls
+// shouldRetry - tells if Action should be rerun
+type fakeRetryProvider struct {
+	ctr         int
+	shouldRetry bool
+}
+
+// IsRetriable returns info if action should be rerun. Every call to this
+// function increments counter.
+func (frp *fakeRetryProvider) IsRetriable(error) bool {
+	frp.ctr++
+	return frp.shouldRetry
+}
+
+func TestRetriableAction(t *testing.T) {
+	for _, tc := range []struct {
+		name             string
+		shouldRetry      bool
+		wantError        bool
+		wantRunThreshold int
+		wantRetries      int
+		wantRun          int
+	}{
+		{
+
+			name:             "should not retry",
+			shouldRetry:      false,
+			wantError:        true,
+			wantRunThreshold: 5,
+			wantRetries:      1,
```

Comment:
nit: this is a little confusing, I would expect `0` here (because there is `4` below). Maybe counting only "retry = true" or changing to `wantRetriableCalls` would help?

Reply:
I will go with `wantRetriableCalls` becaue you need at least 1 call to check if retry is needed.